src/Agent.Listener/Configuration/ConfigurationManager.cs

```diff
@@ -253,6 +253,33 @@ public AgentSettings LoadSettings()
                 }
             }
 
+            // respect the serverUrl resolve by server.
+            // in case of agent configured using collection url instead of account url.
+            string agentServerUrl;
+            if (agent.Properties.TryGetValidatedValue<string>("ServerUrl", out agentServerUrl) &&
+                !string.IsNullOrEmpty(agentServerUrl))
+            {
+                Trace.Info($"Agent server url resolve by server: '{agentServerUrl}'.");
+
+                // we need make sure the Host component of the url remain the same.
+                Uri inputServerUrl = new Uri(serverUrl);
+                Uri serverReturnedServerUrl = new Uri(agentServerUrl);
+                if (!string.Equals(inputServerUrl.GetComponents(UriComponents.Host, UriFormat.Unescaped),
```

Comment:
Uri.Compare(new Uri(""), new Uri(""), UriComponents.Host, UriFormat.Unescaped, StringComparison.OrdinalIgnoreCase)

Seems better to use this method for comparing the host name part.